src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt

```diff
@@ -1 +1,5 @@
 #nullable enable
+~NuGet.Commands.RestoreArgs.NewMappingID.get -> string
+~NuGet.Commands.RestoreArgs.NewMappingID.set -> void
+~NuGet.Commands.RestoreArgs.NewMappingSource.get -> string
+~NuGet.Commands.RestoreArgs.NewMappingSource.set -> void
```

Comment:
This doesn't feel right to me. Merging two different package source mappings doesn't feel like something that Restore should know about.

Take PM UI preview install as a similar example (in fact, this PR modifies the PreviewInstall API signature). it doesn't have separate "actually installed packages" and "new packages to install" inputs. It has just one installed packages list (the data type is PackageSpec), and PM UI modifies the project's actual/current installed packages list with the changes from the PM UI, and then runs a restore, and from RestoreArg's point of view, it looks like every other restore.

Looking at RestoreArgs, it doesn't have the data types needed to override default values, like which packages a project references, or the settings to use. This makes sense, since `RestoreArgs` is a representations of the command line arguments of a restore. `RestoreRunner` has another overload that takes in `RestoreRequest`, which does have the right data types. So, putting aside the layers of abstraction, fundamentally I think that it should 1. Clone settings, and merge the new package source mapping, so it looks like the changes have been written to disk, but it's really just in-memory. 2. Create the appropriate RestoreRequests for the solution, using the custom settings, rather than the "real" settings. I expect that the PreviewInstall already does this to enable package install preview. 3. Call RestoreRunner.RunAsync with the custom RestoreRequests

In fact, since `RestoreArgs` is a representation of the command line args for a restore, it's another reason why this doesn't feel right. This PR/feature isn't adding a way to use different package source mapping config on the command line, so this doesn't seem like the right data structure to modify.

Reply:
> Merging two different package source mappings doesn't feel like something that Restore should know about.

Hm, I see your point. Especially if CLI won't need a feature like "enable automatically mapping newly installed packages to default sources" or something, then CLI would never need to use these args to inject a new mapping.
It's probably trivial, but I'll have to think about refactoring that after vacation.

Appreciate all input in the meantime!

Reply:
I looked over the changes, and I like @zivkan's idea. Beyond that, it seems pretty reasonable what we're trying to do. 

- Adding `packageid` ensures that the new id can come from the selected source.
- Adding `*` in the preview restore allows new packages to come from the selected source only. 

That in addition to the post-restore handling in case of failure/user rejecting, I think this should be pretty good.